### PR TITLE
[fix] Allow URL instances as URL for WebSocket constructors

### PR DESCRIPTION
**Pull request by 3rd-Eden**
Addresses https://github.com/websockets/ws/issues/1143

### CODE REVIEW

File: test/websocket.test.js [hunk 1]
```diff
@@ -5,6 +5,7 @@
 const assert = require('assert');
 const crypto = require('crypto');
 const https = require('https');
+const { URL } = require('url');
```
Comment (lpinca) [inline test/websocket.test.js]: Doesn't work on Node.js 4.

File: lib/websocket.js [hunk 2]
```diff
@@ -463,9 +463,17 @@ function initAsClient (address, protocols, options) {
   }
 
   this._isServer = false;
-  this.url = address;
 
-  const serverUrl = url.parse(address);
+  let serverUrl;
```
Comment (lpinca) [inline lib/websocket.js]: We are still using `var` when `const` can not be used.

Comment (3rd-Eden) [reply lib/websocket.js]: FML, I've been writing to much ES6.

**Comment by lpinca** [comment]
https://github.com/websockets/ws/blob/57d3dbb9e1b6d81132bbf2c18a35f09eab783a87/lib/websocket.js#L513

No longer works when using a WHATWG URL.

File: test/websocket.test.js (hunk 1 above)
Comment (3rd-Eden) [reply test/websocket.test.js]: Node 4 doesn't have the URL instance either, so I've opted for adding in some horrible if statements in the test suite :/

File: lib/websocket.js [hunk 3]
```diff
@@ -511,6 +519,9 @@ function initAsClient (address, protocols, options) {
   }
   if (options.host) requestOptions.headers.Host = options.host;
   if (serverUrl.auth) requestOptions.auth = serverUrl.auth;
+  else if (serverUrl.username && serverUrl.password) {
```
Comment (lpinca) [inline lib/websocket.js]: Still not enough I think:

```
var u = new url.URL('ws://foo@example.org')
undefined
> u
URL {
  href: 'ws://foo@example.org/',
  origin: 'ws://example.org',
  protocol: 'ws:',
  username: 'foo',
  password: '',
  host: 'example.org',
  hostname: 'example.org',
  port: '',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: '' }
```

`auth` should be `'foo'` but it doesn't get set as `serverUrl.password` is falsy.

File: test/websocket.test.js [hunk 4]
```diff
@@ -361,6 +362,15 @@ describe('WebSocket', function () {
 
         assert.strictEqual(ws.url, url);
       });
+
+      if (URL) {
+        it('accepts URL instances as url', function () {
```
Comment (lpinca) [inline test/websocket.test.js]: I would use `this.skip()` for consistency if  WHATWG is not supported.

**Comment by lpinca** [comment]
I think this can be merged.

**Comment by 3rd-Eden** [comment]
yeah, sorry, my internet decided to shit it self today :/

**Pull request merged**